### PR TITLE
Fix for rakudo/rakudo#2178

### PR DESCRIPTION
**Pull request by vrurg**
The fix is done by implementing cooperation between dispatchers. If a
dispatcher finds out that next call in the queue is a invocation handler
with attribute `$!dispatcher` containing a object of descendant type of
`Metamodel::BaseDispatcher` it passes control over to that dispatcher by
calling either `enter_with_args` or `enter_with_capture` method. The
subsequent dispatcher passes control back to the calling dispatcher upon
exhausting its own queue.

The known problems of this fix:

- there is no clear way to determine if next call is a invocation
handler. The approach used by this fix is to wrap the attempt to read
from `$!dispatcher` attribute into a try block is hacky and would have
performance penalty. Must be changed as soon as more straightforward
approach is developed or found.

#2178

### CODE REVIEW

**Comment by vrurg** [comment]
A problem has been discovered which needs fixing before this PR is ready.

**Comment by vrurg** [comment]
The problem has been resolved, PR is ready for review.

File: src/Perl6/Metamodel/Dispatchers.nqp [hunk 1]
```diff
@@ -1,32 +1,64 @@
...
+    has $!next_dispatcher; # The dispatcher we must pass control to when own queue exhausts
 
     method candidates() { @!candidates }
 
-    method exhausted() { $!idx >= +@!candidates }
+    method exhausted() { $!idx >= +@!candidates && (!nqp::isconcrete($!next_dispatcher) || $!next_dispatcher.exhausted()) }
 
     method last()      { @!candidates := [] }
 
+    method set_next_dispatcher ($next_dispatcher) { $!next_dispatcher := $next_dispatcher }
```
Comment (MasterDuke17) [inline src/Perl6/Metamodel/Dispatchers.nqp]: Completely unrelated to what this PR actually does, but you have a couple methods with a space between the method name and the signature.

Comment (vrurg) [reply src/Perl6/Metamodel/Dispatchers.nqp]: My Perl6 habbit. Will fix now. Thanks!

**Pull request merged**